File: recidiviz/tests/case_triage/workflows/utils_test.py
#  Recidiviz - a data platform for criminal justice reform
#  Copyright (C) 2023 Recidiviz, Inc.
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <https://www.gnu.org/licenses/>.
#  =============================================================================
"""Tests for workflows api utils."""
from unittest import TestCase
from unittest.mock import MagicMock, patch

from freezegun import freeze_time

from recidiviz.case_triage.workflows.utils import (
    allowed_twilio_dev_recipient,
    get_sms_request_firestore_path,
)


class TestWorkflowsRouteUtils(TestCase):
    """Test suite for workflows route utils"""

    @patch("recidiviz.case_triage.workflows.utils.FirestoreClientImpl")
    def test_number_allowed(self, mock_firestore_impl: MagicMock) -> None:
        mock_firestore_impl.return_value.get_document.return_value.get.return_value.to_dict.return_value = {
            "allowedTwilioRecipients": {"8884445555": True}
        }

        self.assertTrue(allowed_twilio_dev_recipient("8884445555"))

    @patch("recidiviz.case_triage.workflows.utils.FirestoreClientImpl")
    def test_number_explicitly_disallowed(self, mock_firestore_impl: MagicMock) -> None:
        mock_firestore_impl.return_value.get_document.return_value.get.return_value.to_dict.return_value = {
            "allowedTwilioRecipients": {"5558675309": False}
        }

        self.assertFalse(allowed_twilio_dev_recipient("5558675309"))

    @patch("recidiviz.case_triage.workflows.utils.FirestoreClientImpl")
    def test_number_not_listed(self, mock_firestore_impl: MagicMock) -> None:
        mock_firestore_impl.return_value.get_document.return_value.get.return_value.to_dict.return_value = {
            "allowedTwilioRecipients": {}
        }

        self.assertFalse(allowed_twilio_dev_recipient("8775994444"))

    @patch("recidiviz.case_triage.workflows.utils.FirestoreClientImpl")
    def test_key_missing_from_config(self, mock_firestore_impl: MagicMock) -> None:
        mock_firestore_impl.return_value.get_document.return_value.get.return_value.to_dict.return_value = (
            {}
        )

        self.assertFalse(allowed_twilio_dev_recipient("8775994444"))

    @patch("recidiviz.case_triage.workflows.utils.FirestoreClientImpl")
    def test_firestore_failure_returns_false(
        self, mock_firestore_impl: MagicMock
    ) -> None:
        mock_firestore_impl.return_value.get_document.side_effect = Exception

        self.assertFalse(allowed_twilio_dev_recipient("8775994444"))

    @freeze_time("2023-01-01 01:23:45")
    def test_get_sms_request_firestore_path(self) -> None:
        self.assertEqual(
            get_sms_request_firestore_path(state="US_CA", recipient_external_id="1234"),
            "clientUpdatesV2/us_ca_1234/milestonesMessages/milestones_01_2023",
        )

File: frontends/admin-panel/src/components/StateUserPermissions/__tests__/utils.test.ts
// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2025 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { addWeeks, subWeeks } from "date-fns";

import { FeatureVariants, RouteRecord } from "../../../types";
import {
  checkPermissionsEquivalence,
  mergeFeatureVariants,
  mergeRoutes,
} from "../utils";

const activeDate1 = new Date("2024-04-30T14:45:09.865Z");
const activeDate2 = new Date("2024-05-03T15:37:07.119Z");

describe("mergeRoutes", () => {
  test("merges routes for single role", () => {
    const routesSingleRoleTestData: RouteRecord[] = [
      {
        lantern: true,
        psi: false,
        system_libertyToPrison: false,
        system_prison: false,
      },
    ];

    expect(mergeRoutes(routesSingleRoleTestData)).toEqual({
      lantern: true,
      psi: false,
      system_libertyToPrison: false,
      system_prison: false,
    });
  });

  test("merges routes for multiple roles without conflicts", () => {
    const routesMultipleRolesWithoutConflictsTestData: RouteRecord[] = [
      {
        lantern: true,
        psi: false,
      },
      {
        psi: false,
        system_prison: true,
      },
    ];

    expect(mergeRoutes(routesMultipleRolesWithoutConflictsTestData)).toEqual({
      lantern: true,
      psi: false,
      system_prison: true,
    });
  });

  test("merges routes for multiple roles with conflicts", () => {
    const routesMultipleRolesWithConflictsTestData: RouteRecord[] = [
      {
        lantern: true,
        psi: false,
      },
      {
        lantern: false,
        psi: true,
        system_prison: true,
      },
    ];

    expect(mergeRoutes(routesMultipleRolesWithConflictsTestData)).toEqual({
      lantern: true,
      psi: true,
      system_prison: true,
    });
  });
});

describe("mergeFeatureVariants", () => {
  test("merges feature variants for single role", () => {
    const featureVariantsSingleRoleTestData: FeatureVariants[] = [
      { feature1: {}, feature2: { activeDate: activeDate1 } },
    ];

    expect(mergeFeatureVariants(featureVariantsSingleRoleTestData)).toEqual({
      feature1: {},
      feature2: { activeDate: activeDate1 },
    });
  });

  test("merges feature variants for multiple roles without conflicts", () => {
    const featureVariantsMultipleRolesNoConflictTestData: FeatureVariants[] = [
      { feature1: false },
      { feature2: {} },
    ];

    expect(
      mergeFeatureVariants(featureVariantsMultipleRolesNoConflictTestData)
    ).toEqual({ feature1: false, feature2: {} });
  });

  test("merges feature variants for multiple roles with conflicts", () => {
    const featureVariantsMultipleRolesWithConflictsTestData: FeatureVariants[] =
      [
        {
          feature1: false,
          feature2: { activeDate: activeDate1 },
          feature3: false,
          feature4: { activeDate: activeDate2 },
          feature5: {},
          feature6: { activeDate: activeDate1 },
        },
        {
          feature1: {},
          feature2: {},
          feature3: { activeDate: activeDate2 },
          feature4: { activeDate: activeDate1 },
          feature5: { activeDate: activeDate2 },
          feature6: false,
        },
      ];

    expect(
      mergeFeatureVariants(featureVariantsMultipleRolesWithConflictsTestData)
    ).toEqual({
      feature1: {},
      feature2: {},
      feature3: { activeDate: activeDate2 },
      feature4: { activeDate: activeDate1 },
      feature5: {},
      feature6: { activeDate: activeDate1 },
    });
  });
});

describe("checkPermissionsEquivalence", () => {
  test("equivalent when default permissions are always on", () => {
    expect(checkPermissionsEquivalence({}, {})).toBe(true);
  });

  test("equivalent when default permissions have a past activeDate", () => {
    const defaultPermissionsPastDate = new Date("2025-01-01");
    const userPermissionsPastDate = subWeeks(defaultPermissionsPastDate, 1);
    expect(
      checkPermissionsEquivalence(
        { activeDate: defaultPermissionsPastDate },
        {}
      )
    ).toBe(true);
    expect(
      checkPermissionsEquivalence(
        { activeDate: defaultPermissionsPastDate },
        { activeDate: userPermissionsPastDate }
      )
    ).toBe(true);
  });

  test("equivalent when both default and user permissions have a future activeDate", () => {
    const defaultPermissionsFutureDate = addWeeks(new Date(), 1);
    const userPermissionsFutureDate = addWeeks(new Date(), 2);
    expect(
      checkPermissionsEquivalence(
        { activeDate: defaultPermissionsFutureDate },
        { activeDate: userPermissionsFutureDate }
      )
    ).toBe(true);
  });

  test("not equivalent when default permissions have a future date and user permissions are always on", () => {
    const defaultPermissionsFutureDate = addWeeks(new Date(), 1);
    expect(
      checkPermissionsEquivalence(
        { activeDate: defaultPermissionsFutureDate },
        {}
      )
    ).toBe(false);
  });

  test("not equivalent when default permissions have a future date and user permissions have a past date", () => {
    const defaultPermissionsFutureDate = addWeeks(new Date(), 1);
    const userPermissionsPastDate = subWeeks(new Date(), 1);
    expect(
      checkPermissionsEquivalence(
        { activeDate: defaultPermissionsFutureDate },
        { activeDate: userPermissionsPastDate }
      )
    ).toBe(false);
  });

  test("equivalent when default permissions are true", () => {
    expect(checkPermissionsEquivalence(true, true)).toBe(true);
  });

  test("equivalent when both default and user permissions are false", () => {
    expect(checkPermissionsEquivalence(false, false)).toBe(true);
  });

  test("not equivalent when default permissions are false and user permissions are true", () => {
    expect(checkPermissionsEquivalence(false, true)).toBe(false);
  });
});
